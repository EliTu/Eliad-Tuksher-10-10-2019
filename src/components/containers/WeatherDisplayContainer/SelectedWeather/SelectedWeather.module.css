@import '../../../../variables.css';

.SelectedWeatherStyles {
    width: 80%;
    height: 40%;
    margin: 0 auto;
    padding: 0 2rem;
    display: flex;
    justify-content: center;
    align-items: center;
    justify-items: center;
}

.buttonWrapper {
    width: 30%;
}

.SelectedWeatherStyles ul {
    width: 50%;
    padding: 0;
}

.SelectedWeatherStyles li {
    list-style: none;
    font-size: var(--text-size-extra-large);
    margin: 0 auto;
    margin-bottom: 1rem;
    text-align: center;
    font-weight: 800;
    align-self: center;
}

.SelectedWeatherStyles li:nth-child(2) {
    font-size: var(--text-size-huge);
}

.SelectedWeatherStyles button {
    border: none;
    height: 10px;
    background-color: rgba(255, 0, 0, 0);
    margin: 2rem;
    padding: 0;
    outline: none;
    align-self: start;
}