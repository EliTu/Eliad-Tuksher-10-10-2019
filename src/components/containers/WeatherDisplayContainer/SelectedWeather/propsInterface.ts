export interface IProps {
	setNewFavoriteItem: (
		key: string,
		cityName: string,
		countryName: string
	) => void;
	removeFromFavorites: (key: string) => void;
	favoritesList: string[];
	weatherData: {
		LocalObservationDateTime: string;
		EpochTime: number;
		WeatherText: string;
		WeatherIcon: number;
		IsDayTime: boolean;
		Link: string;
		Temperature: {
			Metric: {
				Value: number;
				Unit: string;
			};
		};
		cityName: string;
		countryName: string;
		key: string;
	};
	isLoading: boolean;
	initFavorites: () => void;
}
