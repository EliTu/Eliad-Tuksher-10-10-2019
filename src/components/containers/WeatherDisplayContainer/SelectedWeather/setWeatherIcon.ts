import {
	faSun,
	faCloudSun,
	faCloud,
	faCloudRain,
	faCloudShowersHeavy,
	faCloudSunRain,
	faWind,
	faSmog,
	faCloudMoon,
	faCloudMoonRain,
	faMoon,
	faSnowflake,
	IconDefinition,
} from '@fortawesome/free-solid-svg-icons';

const setWeatherIcon = (isDay: boolean, iconNumber: number) => {
	let icon: IconDefinition;

	switch (iconNumber) {
		case 1:
		case 2:
		case 3:
		case 33:
		case 34:
			icon = isDay ? faSun : faMoon;
			break;

		case 5:
		case 21:
		case 36:
			icon = isDay ? faCloudSun : faCloudMoon;
			break;

		case 6:
		case 20:
		case 7:
		case 8:
		case 35:
		case 38:
			icon = isDay ? faCloud : faCloudMoon;
			break;

		case 11:
			icon = faSmog;
			break;

		case 18:
		case 19:
		case 25:
		case 26:
		case 29:
		case 39:
		case 40:
			icon = isDay ? faCloudRain : faCloudMoonRain;
			break;

		case 14:
		case 17:
		case 16:
			icon = faCloudSunRain;
			break;

		case 15:
			icon = faCloudShowersHeavy;
			break;

		case 32:
			icon = faWind;
			break;

		case 22:
		case 24:
		case 23:
			icon = faSnowflake;
			break;

		default:
			icon = isDay ? faCloudSun : faCloudMoon;
			break;
	}
	return icon;
};

export default setWeatherIcon;
