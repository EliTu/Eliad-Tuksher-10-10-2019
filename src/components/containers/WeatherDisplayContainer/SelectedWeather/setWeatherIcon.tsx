import React from 'react';
import Icon from '../../../display/UI/Icon/Icon';
import {
	faSun,
	faCloudSun,
	faCloud,
	faCloudRain,
	faCloudShowersHeavy,
	faCloudSunRain,
	faWind,
	faSmog,
	faCloudMoon,
	faCloudMoonRain,
	faMoon,
	faSnowflake,
} from '@fortawesome/free-solid-svg-icons';

const setWeatherIcon = (isDay: boolean, weatherText: string) => {
	let icon: JSX.Element;
	console.log(isDay);
	console.log(weatherText);

	switch (weatherText) {
		case 'Sunny':
		case 'Mostly Sunny':
		case 'Partly Sunny w/ Flurries':
		case 'Clear':
		case 'Mostly Clear':
			icon = isDay ? (
				<Icon iconType={faSun} size={'7x'}></Icon>
			) : (
				<Icon iconType={faMoon} size={'7x'}></Icon>
			);
			break;

		case 'Hazy Sunshine':
		case 'Partly Sunny':
		case 'Intermittent Clouds':
			icon = isDay ? (
				<Icon iconType={faCloudSun} size={'7x'}></Icon>
			) : (
				<Icon iconType={faCloudMoon} size={'7x'}></Icon>
			);
			break;

		case 'Mostly Cloudy':
		case 'Mostly Cloudy w/ Flurries':
		case 'Cloudy':
		case 'Dreary (Overcast)':
		case 'Partly Cloudy':
			icon = isDay ? (
				<Icon iconType={faCloud} size={'7x'}></Icon>
			) : (
				<Icon iconType={faCloudMoon} size={'7x'}></Icon>
			);
			break;

		case 'Fog':
			icon = <Icon iconType={faSmog} size={'7x'}></Icon>;
			break;

		case 'Rain':
		case 'Sleet':
		case 'Freezing Rain':
		case 'Rain and Snow':
		case 'Mostly Cloudy w/ Showers':
			icon = isDay ? (
				<Icon iconType={faCloudRain} size={'7x'}></Icon>
			) : (
				<Icon iconType={faCloudMoonRain} size={'7x'}></Icon>
			);
			break;

		case 'Partly Sunny w/ Showers':
		case 'Partly Sunny w/ T-Storms':
		case 'Mostly Cloudy w/ T-Storms':
			icon = <Icon iconType={faCloudSunRain} size={'7x'}></Icon>;
			break;

		case 'T-Storms':
			icon = <Icon iconType={faCloudShowersHeavy} size={'7x'}></Icon>;
			break;

		case 'Windy':
			icon = <Icon iconType={faWind} size={'7x'}></Icon>;
			break;

		case 'Snow':
		case 'Ice':
		case 'Mostly Cloudy w/ Snow':
			icon = <Icon iconType={faSnowflake} size={'7x'}></Icon>;
			break;

		default:
			icon = isDay ? (
				<Icon iconType={faCloudSun} size={'7x'}></Icon>
			) : (
				<Icon iconType={faCloudMoon} size={'7x'}></Icon>
			);
			break;
	}

	return icon;
};

export default setWeatherIcon;
