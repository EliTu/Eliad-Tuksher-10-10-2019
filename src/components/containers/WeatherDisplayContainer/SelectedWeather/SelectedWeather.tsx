import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import SelectedWeatherInfo from './SelectedWeatherInfo/SelectedWeatherInfo';
import FavIcon from '../../../display/UI/Icon/FavIcon/FavIcon';
import { fireCurrentWeatherHttpRequest } from '../store/actions';
import {
	addToFavoritesAction,
	removeFromFavoritesAction,
	initFavoritesAction,
} from '../../Favorites/store/actions';
import {
	setCorrectDateFormat,
	setDayOfTheWeek,
} from '../../../../utilities/convert-functions/dates';
import setWeatherIcon from './setWeatherIcon';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import styles from './SelectedWeather.module.css';

interface IProps {
	setNewFavoriteItem: (
		key: string,
		cityName: string,
		countryName: string
	) => void;
	removeFromFavorites: (key: string) => void;
	favoritesList: string[];
	weatherData: {
		LocalObservationDateTime: string;
		EpochTime: number;
		WeatherText: string;
		WeatherIcon: number;
		IsDayTime: boolean;
		Link: string;
		Temperature: {
			Metric: {
				Value: number;
				Unit: string;
			};
		};
		cityName: string;
		countryName: string;
		key: string;
	};
	isLoading: boolean;
	initFavorites: () => void;
}

export const SelectedWeather: React.FC<IProps> = ({
	setNewFavoriteItem,
	removeFromFavorites,
	favoritesList,
	weatherData,
	isLoading,
	initFavorites,
}) => {
	const { SelectedWeatherStyles, buttonWrapper } = styles;

	const [isFavorite, setIsFavorite] = useState(false);
	const [weatherIconType, setWeatherIconType] = useState();

	// Destructured weatherData:
	const {
		LocalObservationDateTime: localDate,
		EpochTime,
		WeatherText,
		WeatherIcon,
		IsDayTime,
		Link,
		Temperature,
		cityName,
		countryName,
		key,
	} = weatherData;

	// Set correct date formats:

	const day = !isLoading && EpochTime ? setDayOfTheWeek(EpochTime) : '';
	const date = !isLoading && localDate ? setCorrectDateFormat(localDate) : '';

	useEffect(() => {
		initFavorites();
	}, [initFavorites]);

	useEffect(() => {
		const checkForFavoriteListing = (): any => {
			let isListed;
			if (favoritesList)
				isListed = favoritesList.some((el: any) => el.key === key);

			return isListed;
		};
		setIsFavorite(() => checkForFavoriteListing());
	}, [favoritesList, key, isFavorite]);

	// Set the current weather Icon according to data:
	useEffect(() => {
		let weatherIcon: IconDefinition;
		if (!isLoading && WeatherIcon)
			weatherIcon = setWeatherIcon(IsDayTime, WeatherIcon);
		setWeatherIconType(() => weatherIcon);
	}, [IsDayTime, WeatherIcon, isLoading]);

	const handleFavoriteButtonClick = (key: string) => {
		if (!isFavorite) {
			setNewFavoriteItem(key, cityName, countryName);
			setIsFavorite(() => true);
		} else {
			removeFromFavorites(key);
			setIsFavorite(() => false);
		}
	};

	return (
		<div className={SelectedWeatherStyles}>
			{!isLoading && weatherData.key && (
				<>
					<SelectedWeatherInfo
						weatherIconType={weatherIconType}
						isLoading={isLoading}
						infoLink={Link}
						cityName={cityName}
						countryName={countryName}
						temperature={Temperature}
						date={date}
						day={day}
					/>
					<ul>
						<li>{WeatherText}</li>
						<li>
							{Temperature &&
								`${Temperature.Metric.Value}${Temperature.Metric.Unit}`}
						</li>
					</ul>
					<div className={buttonWrapper}>
						<button onClick={() => handleFavoriteButtonClick(key)}>
							<FavIcon isFavorite={isFavorite} />
						</button>
					</div>
				</>
			)}
		</div>
	);
};

const mapStateToProps = (state: any) => {
	return {
		weatherData: state.currentWeather.selectedResult,
		isLoading: state.currentWeather.isLoading,
		searchResults: state.search.results,
		favoritesList: state.favorites.favoritesList,
	};
};

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => {
	return {

		setNewFavoriteItem: (
			key: string,
			cityName: string,
			countryName: string
		) => dispatch(addToFavoritesAction(key, cityName, countryName)),

		removeFromFavorites: (key: string) =>
			dispatch(removeFromFavoritesAction(key)),
		initFavorites: () => dispatch(initFavoritesAction()),
	};
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(SelectedWeather);
