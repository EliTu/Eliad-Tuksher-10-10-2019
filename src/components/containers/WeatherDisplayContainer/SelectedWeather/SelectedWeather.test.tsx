import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { SelectedWeather } from './SelectedWeather';
import initialShallowRender from '../../../../utilities/test-utilities/initialShallowRender';

describe('SelectedWeather Component', () => {
	let component: ShallowWrapper;
	beforeEach(
		() =>
			(component = shallow(
				<SelectedWeather
					currentWeatherHttpRequest={() => {}}
					weatherData={{
						WeatherText: 'abc',
						isDayTime: false,
						Link: 'www',
						Temperature: { Metric: { Value: 20, Unit: 'C' } },
					}}
					isLoading={false}
					searchResults={{ localizedName: 'abc' }}
				/>
			))
	);

	it('should render without errors', () =>
		initialShallowRender(component, '.SelectedWeatherStyles'));

	it('should render a div element with the data info if isLoading is false', () => {
		expect(component.children().find('div')).toBeTruthy();
		expect(component.children().children().length).toBe(4);
	});

	it('should render an Icon component if isLoading is false', () => {
		const icon = component
			.children()
			.children()
			.at(0);

		expect(icon.length).toBe(1);
	});

	it('should render a ul component with 2 li elements', () => {
		const ul = component
			.children()
			.children()
			.at(1);
		const li = ul.children();

		expect(ul.length).toBe(1);
		expect(li.length).toBe(2);
	});

	it('should render a p tag with a text', () => {
		const pTag = component.find('p');

		expect(pTag.length).toBe(1);
		expect(pTag.text()).toBe('abc');
	});

	it('should have a button with a FavIcon component inside', () => {
		const button = component.find('button');
		const favIcon = button.children();

		expect(button.length).toBe(1);
		expect(favIcon.length).toBe(1);
	});
});
