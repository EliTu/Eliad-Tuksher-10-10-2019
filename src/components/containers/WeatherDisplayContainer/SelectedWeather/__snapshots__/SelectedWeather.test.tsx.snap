// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedWeather Component should render without errors 1`] = `
<div
  className="SelectedWeatherStyles"
>
  <div
    className="WeatherInfo"
  >
    <Icon
      size="7x"
    />
    <ul>
      <li>
        <a
          href="www"
        >
          Tel-Aviv
        </a>
      </li>
      <li>
        20C
      </li>
    </ul>
  </div>
  <p>
    abc
  </p>
  <button>
    <FavIcon
      isFavorite={false}
    />
  </button>
</div>
`;
